# IN PROGRESS
# ------------------------------------------------ #
# 2bRAD Trim and QC Snakefile
# Kirsten Carlson
# Updated 1/2026
# ------------------------------------------------ #

# This Snakefile runs the pipeline for processing 2bRAD data.

# Scripts needed for this pipeline:
# From https://github.com/z0on/2bRAD_denovo
#  - trim2bRAD_2barcodes_dedup.pl
#  - uniquerOne.pl
#  - mergeUniq.pl
#  - concatFasta.pl

# ------------------------------------------------ #
# CONFIGURATION
# ------------------------------------------------ #

configfile: "../config/config.yaml"
# PRJNA accession number
prjna = config["prjna"]
# Conda environment
env: config["env"]
# Raw fastq file directory
rawfq_dir = config["rawfq_dir"] 
# Raw QC file output directory
rawqc_dir = config["rawqc_dir"]  
# Trimmed fastq file directory
trimfq_dir = config["trimfq_dir"] 
# Trimmed QC file output directory
trimqc_dir = config["trimqc_dir"] 
# Excluded symbiont trimmed fastq directory
trimfq_noSymb_dir = config["trimfq_noSymb_dir"] 
# Merged unique reads directory
merge_dir = config["merge_dir"] 
# Aligned symbiont bam directory
symb_align_dir = config["symb_align_dir"]
# Kraken database directory
kraken_db_std = config["kraken_db_std"]
# De novo reference genome directory
denovo_ref_dir = config["denovo_ref_dir"]
# Script directory
scripts_dir = config["scripts_dir"] 
# Symbiont genomes
symbiont_genomes = config["symbiont_genomes"]
# Symbiont concatenated reference genome
reference_symbiont = config["reference_symbiont"] 
# Script directory
scripts_dir = config["scripts_dir"]
# Number of threads to use
threads = config["threads"] 
quality = config["quality"]
minlen = config["minlen"] 

# ------------------------------------------------ #
# LOCA RULE FILES
# ------------------------------------------------ #
include: "rules/metadata.smk"
include: "rules/download.smk"
include: "rules/trim_qc.smk"
include: "rules/de_novo_symb_pipeline.smk"

import pandas as pd

# Read SRR list to define final targets
srr_df = pd.read_csv(f"{config['resource_dir']}/srr_2brad_list.txt", sep="\t", header=None, names=["SRR", "SAMPLEID"])
SAMPLES = srr_df["SAMPLEID"].tolist()  

# Define final target(s)
rule all:
    input:
        expand(f"{denovo_ref_dir}/cleanup.done", sample=SAMPLES)
